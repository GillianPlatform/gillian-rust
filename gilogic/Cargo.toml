[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }