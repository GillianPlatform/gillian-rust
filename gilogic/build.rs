fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
