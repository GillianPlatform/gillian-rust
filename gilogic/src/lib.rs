#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
