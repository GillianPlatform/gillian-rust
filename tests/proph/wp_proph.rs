//@check-pass
#![feature(concat_idents)]

extern crate gilogic;

use gilogic::{
    macros::{
        assertion, extract_lemma, lemma, predicate, prophecies::with_freeze_lemma_for_mutref,
        specification,
    },
    mutref_auto_resolve,
    prophecies::{Ownable, Prophecised, Prophecy},
};

struct WP<T> {
    x: *mut N<T>,
    y: *mut N<T>,
}

struct N<T> {
    v: T,
    o: *mut N<T>,
}

#[predicate]
fn wp<T: Ownable>(
    wp: In<WP<T>>,
    x: *mut N<T>,
    y: *mut N<T>,
    model: (T::RepresentationTy, T::RepresentationTy),
) {
    assertion!(|v_x, v_y, v_x_m, v_y_m|
        (wp == WP { x, y }) *
        (x -> N { v: v_x, o: y }) *
        (y -> N { v: v_y, o: x }) *
        v_x.own(v_x_m) * v_y.own(v_y_m) *
        (model == (v_x_m, v_y_m))
    )
}

#[with_freeze_lemma_for_mutref(
    lemma_name = freeze_xy,
    predicate_name = wp_ref_mut_xy,
    frozen_variables = [x, y],
    inner_predicate_name = wp_ref_mut_inner_xy
)]
impl<T: Ownable> Ownable for WP<T> {
    type RepresentationTy = (T::RepresentationTy, T::RepresentationTy);

    #[predicate]
    fn own(self, model: Self::RepresentationTy) {
        assertion!(|x: *mut N<T>, y: *mut N<T>| wp(self, x, y, model))
    }
}

// TODO: I need to not create matching plans for lemmas that do not have a proof!
// Because the prophecy variable in the post condition is just generated out of nowhere.
#[extract_lemma(
    forall p, x: *mut N<T>, y: *mut N<T>.
    model m.
    extract model mx: (T::RepresentationTy, T::RepresentationTy).
    from { wp_ref_mut_xy(p, m, x, y) }
    extract { Ownable::own(&mut (*x).v, mx) }
    prophecise { (mx, m.1) }
)]
fn extract_x<'a, T: Ownable>(p: &'a mut WP<T>) -> Prophecy<T::RepresentationTy>;

//
// fn extract_tuple<T>(p : &mut (T, T)) -> &mut T;

impl<T: Ownable> WP<T> {
    #[specification(
        forall lx, ly.
        requires { x.own(lx) * y.own(ly) }
        exists ret_v.
        ensures { ret.own(ret_v) * $ret_v == (lx, ly)$ }
    )]
    fn new(x: T, y: T) -> Self {
        let null: *mut N<T> = std::ptr::null_mut();

        let xbox = Box::new(N { v: x, o: null });
        let ybox = Box::new(N { v: y, o: null });

        let xptr = Box::leak(xbox) as *mut N<T>;
        let yptr = Box::leak(ybox) as *mut N<T>;

        unsafe {
            (*yptr).o = xptr;
            (*xptr).o = yptr;
        }
        WP { x: xptr, y: yptr }
    }

    #[specification(
        forall current, proph, xm.
        requires { self.own((current, proph)) * x.own(xm) }
        ensures { $proph == (xm, current.1)$ }
    )]
    fn assign_first(&mut self, x: T) {
        unsafe {
            (*self.x).v = x;
            mutref_auto_resolve!(self);
        }
    }

    #[specification(forall cself, pself.
        requires { self.own((cself, pself)) }
        exists c, p.
        ensures {  ret.own((c, p)) * $cself.1 == pself.1 && pself.0 == p && cself.0 == c$ }
    )]
    fn first_mut<'a>(&'a mut self) -> &'a mut T {
        unsafe {
            freeze_xy(self);
            let ret = &mut (*self.x).v;
            let proph = extract_x(self);
            ret.with_prophecy(proph)
        }
    }
}
