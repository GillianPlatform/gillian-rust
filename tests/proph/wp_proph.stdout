#begin_init_data
{"Global":["Struct","ReprRust",[]],"N":["Struct","ReprRust",[["v",["Param",0]],["o",["Ptr",{"mut":true,"ty":["Adt",["N",[["Param",0]]]]}]]]],"WP":["Struct","ReprRust",[["x",["Ptr",{"mut":true,"ty":["Adt",["N",[["Param",0]]]]}]],["y",["Ptr",{"mut":true,"ty":["Adt",["N",[["Param",0]]]]}]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_pcy.gil", "i__lang_pcy.gil";

axiomatic spec "<&mut WP<T> as gilogic::prophecies::Prophecised>::prophecy_resolve"(lft, pty_T0, mut_ref)
[[ "<&mut WP<T> as gilogic::prophecies::Ownable>::own"(lft, pty_T0, mut_ref, {{ #current, #future }}) ]]
[[<observation>((#current = #future); )]]
normal

pred "wp::<T>"(+pty_T0, +wp, x, y, model):
  (#wp == {{ #x, #y }})
  * <value>(l-nth(#x, 0i), l-nth(#x, 1i), {{ "adt", "N", {{ pty_T0 }} }}; {{ #v_x, #y }})
  * <value>(l-nth(#y, 0i), l-nth(#y, 1i), {{ "adt", "N", {{ pty_T0 }} }}; {{ #v_y, #x }})
  * "<T as gilogic::prophecies::Ownable>::own"(pty_T0, #v_x, #v_x_m)
  * "<T as gilogic::prophecies::Ownable>::own"(pty_T0, #v_y, #v_y_m)
  * (#model == {{ #v_x_m, #v_y_m }})
  * types(#lvar_0: Obj, #lvar_1: List) * (#x == {{ #lvar_0, #lvar_1 }})
  * types(#lvar_2: Obj, #lvar_3: List) * (#y == {{ #lvar_2, #lvar_3 }})
  * (model == #model)
  * (y == #y)
  * (x == #x)
  * (wp == #wp);


pred "wp_ref_mut_inner_xy2::<T>"(+pLft_0, +pty_T0, +REFERENCE, x, y):
  "wp::<T>"(pty_T0, #SELF, #x, #y, #model)
  * <value>(l-nth(l-nth(#REFERENCE, 0i), 0i), l-nth(l-nth(#REFERENCE, 0i), 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #SELF)
  * <pcy_controller>(l-nth(#REFERENCE, 1i); #model)
  * types(#lvar_4: Obj, #lvar_5: List, #lvar_6: Obj)
  * (#REFERENCE == {{ {{ #lvar_4, #lvar_5 }}, #lvar_6 }})
  * types(#lvar_7: Obj, #lvar_8: List) * (#x == {{ #lvar_7, #lvar_8 }})
  * types(#lvar_9: Obj, #lvar_10: List) * (#y == {{ #lvar_9, #lvar_10 }})
  * (y == #y)
  * (x == #x)
  * (REFERENCE == #REFERENCE);
guard: <lft>(pLft_0; true);

pred "wp_ref_mut_xy2::<T>"(+pLft_gil, +pty_T0, +REFERENCE, model, x, y):
  (#model == {{ #current, #lvar_11 }})
  * <value_observer>(l-nth(#REFERENCE, 1i); #current)
  * "wp_ref_mut_inner_xy2::<T>"(pLft_gil, pty_T0, #REFERENCE, #x, #y)
  * <pcy_value>(l-nth(#REFERENCE, 1i); #lvar_11)
  * types(#lvar_12: Obj, #lvar_13: List, #lvar_14: Obj)
  * (#REFERENCE == {{ {{ #lvar_12, #lvar_13 }}, #lvar_14 }})
  * types(#lvar_15: Obj, #lvar_16: List) * (#x == {{ #lvar_15, #lvar_16 }})
  * types(#lvar_17: Obj, #lvar_18: List) * (#y == {{ #lvar_17, #lvar_18 }})
  * (y == #y)
  * (x == #x)
  * (model == #model)
  * (REFERENCE == #REFERENCE);


pred "<WP<T> as gilogic::prophecies::Ownable>::own"(+pty_T0, +self, model):
  "wp::<T>"(pty_T0, #self, #x, #y, #model) * (model == #model)
  * (self == #self);


abstract pred "<T as gilogic::prophecies::Ownable>::own"(+pty_T0, +self, model);


pred "<&mut WP<T> as gilogic::prophecies::Ownable>::own"(+pLft_erased, +pty_T0, +self, model):
  (#model == {{ #a, #b }})
  * "gilogic::prophecies::mut_ref_inner_proph::<WP<T>>"(pLft_erased, pty_T0, #self)
  * <value_observer>(l-nth(#self, 1i); #a)
  * (#lvar_0 == #b)
  * <pcy_value>(l-nth(#self, 1i); #lvar_0)
  * types(#lvar_1: Obj, #lvar_2: List, #lvar_3: Obj)
  * (#self == {{ {{ #lvar_1, #lvar_2 }}, #lvar_3 }})
  * (model == #model)
  * (self == #self);


pred "<&mut T as gilogic::prophecies::Ownable>::own"(+pLft_erased, +pty_T0, +self, model):
  (#model == {{ #a, #b }})
  * "gilogic::prophecies::mut_ref_inner_proph::<T>"(pLft_erased, pty_T0, #self)
  * <value_observer>(l-nth(#self, 1i); #a)
  * (#lvar_0 == #b)
  * <pcy_value>(l-nth(#self, 1i); #lvar_0)
  * types(#lvar_1: Obj, #lvar_2: List, #lvar_3: Obj)
  * (#self == {{ {{ #lvar_1, #lvar_2 }}, #lvar_3 }})
  * (model == #model)
  * (self == #self);


pred "gilogic::prophecies::mut_ref_inner_proph::<WP<T>>"(+pLft_a, +pty_T0, +this):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #v)
  * "<WP<T> as gilogic::prophecies::Ownable>::own"(pty_T0, #v, #repr)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (this == #this);
guard: <lft>(pLft_a; true);

pred "gilogic::prophecies::mut_ref_inner_proph::<T>"(+pLft_a, +pty_T0, +this):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), pty_T0; #v)
  * "<T as gilogic::prophecies::Ownable>::own"(pty_T0, #v, #repr)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (this == #this);
guard: <lft>(pLft_a; true);

pred "wp_ref_mut_inner_xy2::<T>$$inner"(+pty_T0, +REFERENCE, x, y):
  "wp::<T>"(pty_T0, #SELF, #x, #y, #model)
  * <value>(l-nth(l-nth(#REFERENCE, 0i), 0i), l-nth(l-nth(#REFERENCE, 0i), 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #SELF)
  * <pcy_controller>(l-nth(#REFERENCE, 1i); #model)
  * types(#lvar_4: Obj, #lvar_5: List, #lvar_6: Obj)
  * (#REFERENCE == {{ {{ #lvar_4, #lvar_5 }}, #lvar_6 }})
  * types(#lvar_7: Obj, #lvar_8: List) * (#x == {{ #lvar_7, #lvar_8 }})
  * types(#lvar_9: Obj, #lvar_10: List) * (#y == {{ #lvar_9, #lvar_10 }})
  * (y == #y)
  * (x == #x)
  * (REFERENCE == #REFERENCE);


pred "gilogic::prophecies::mut_ref_inner_proph::<WP<T>>$$inner"(+pty_T0, +this):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #v)
  * "<WP<T> as gilogic::prophecies::Ownable>::own"(pty_T0, #v, #repr)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (this == #this);


pred "gilogic::prophecies::mut_ref_inner_proph::<T>$$inner"(+pty_T0, +this):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), pty_T0; #v)
  * "<T as gilogic::prophecies::Ownable>::own"(pty_T0, #v, #repr)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (this == #this);


lemma "freeze_xy::<T>"(pLft_0, pty_T0, REFERENCE)[[ "<&mut WP<T> as gilogic::prophecies::Ownable>::own"(#pLft_0, #pty_T0, #REFERENCE, #MODEL) * <lft>(#pLft_0; true) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#REFERENCE == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (MODEL == #MODEL) * (REFERENCE == #REFERENCE) * (pty_T0 == #pty_T0) * (pLft_0 == #pLft_0) ]]
[["wp_ref_mut_xy2::<T>"(#pLft_0, #pty_T0, #REFERENCE, #MODEL, #x, #y) * <lft>(#pLft_0; true)]]


spec "WP::<T>::new"(pty_T0, x, y, lx, ly)
[[ "<T as gilogic::prophecies::Ownable>::own"(#pty_T0, #x, #lx) * "<T as gilogic::prophecies::Ownable>::own"(#pty_T0, #y, #ly) * (ly == #ly) * (lx == #lx) * (y == #y) * (x == #x) * (pty_T0 == #pty_T0) ]]
[["<WP<T> as gilogic::prophecies::Ownable>::own"(#pty_T0, ret, #ret_v) * <observation>((#ret_v = {{ #lx, #ly }}); )]]
normal

proc "WP::<T>::new"(pty_T0, x, y) {
        m_1x := x;
        m_2y := y;
        m_12xptr := [alloc]({{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        m_13 := [alloc]({{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        m_15yptr := [alloc]({{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        m_16 := [alloc]({{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
  bb0:  g_0 := "std::ptr::null_mut"({{ "adt", "N", {{ pty_T0 }} }});
        m_3null := g_0;
        goto bb1;
  bb1:  g_1 := m_1x;
        m_6 := g_1;
        m_7 := m_3null;
        g_2 := m_6;
        g_3 := m_7;
        m_5 := {{ g_2, g_3 }};
        goto bb2;
  bb2:  g_4 := m_5;
        g_5 := "std::boxed::Box::<T>::new"({{ "adt", "N", {{ pty_T0 }} }}, g_4);
        m_4xbox := g_5;
        goto bb3;
  bb3:  g_6 := m_2y;
        m_10 := g_6;
        m_11 := m_3null;
        g_7 := m_10;
        g_8 := m_11;
        m_9 := {{ g_7, g_8 }};
        goto bb4;
  bb4:  g_9 := m_9;
        g_10 := "std::boxed::Box::<T>::new"({{ "adt", "N", {{ pty_T0 }} }}, g_9);
        m_8ybox := g_10;
        goto bb5;
  bb5:  g_11 := m_4xbox;
        m_14 := g_11;
        g_12 := [new_lft]();
        pLft_4 := l-nth(g_12, 0i);
        g_13 := m_14;
        g_14 := "std::boxed::Box::<T, A>::leak"(pLft_4, {{ "adt", "N", {{ pty_T0 }} }}, {{ "adt", "Global", {{  }} }}, g_13);
        u := [store_value](l-nth(m_13, 0i), l-nth(m_13, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_14);
        goto bb6;
  bb6:  g_16 := [load_value](l-nth(m_13, 0i), l-nth(m_13, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_15 := l-nth(g_16, 0i);
        u := [store_value](l-nth(m_12xptr, 0i), l-nth(m_12xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, l-nth(g_15, 0i));
        g_17 := m_8ybox;
        m_17 := g_17;
        g_18 := [new_lft]();
        pLft_5 := l-nth(g_18, 0i);
        g_19 := m_17;
        g_20 := "std::boxed::Box::<T, A>::leak"(pLft_5, {{ "adt", "N", {{ pty_T0 }} }}, {{ "adt", "Global", {{  }} }}, g_19);
        u := [store_value](l-nth(m_16, 0i), l-nth(m_16, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_20);
        goto bb7;
  bb7:  g_22 := [load_value](l-nth(m_16, 0i), l-nth(m_16, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_21 := l-nth(g_22, 0i);
        u := [store_value](l-nth(m_15yptr, 0i), l-nth(m_15yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, l-nth(g_21, 0i));
        g_24 := [load_value](l-nth(m_12xptr, 0i), l-nth(m_12xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_23 := l-nth(g_24, 0i);
        m_19 := g_23;
        g_25 := m_19;
        g_27 := [load_value](l-nth(m_15yptr, 0i), l-nth(m_15yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_26 := l-nth(g_27, 0i);
        u := [store_value](l-nth(g_26, 0i), l+ (l-nth(g_26, 1i), {{ {{ "f", 1i, {{ "adt", "N", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_25);
        g_29 := [load_value](l-nth(m_15yptr, 0i), l-nth(m_15yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_28 := l-nth(g_29, 0i);
        m_20 := g_28;
        g_30 := m_20;
        g_32 := [load_value](l-nth(m_12xptr, 0i), l-nth(m_12xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_31 := l-nth(g_32, 0i);
        u := [store_value](l-nth(g_31, 0i), l+ (l-nth(g_31, 1i), {{ {{ "f", 1i, {{ "adt", "N", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_30);
        m_18 := {{  }};
        g_34 := [load_value](l-nth(m_12xptr, 0i), l-nth(m_12xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_33 := l-nth(g_34, 0i);
        m_21 := g_33;
        g_36 := [load_value](l-nth(m_15yptr, 0i), l-nth(m_15yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_35 := l-nth(g_36, 0i);
        m_22 := g_35;
        g_37 := m_21;
        g_38 := m_22;
        m_0 := {{ g_37, g_38 }};
        goto bb8;
  bb8:  goto bb9;
  bb9:  goto bb10;
  bb10: goto bb11;
  bb11: g_39 := m_0;
        ret := g_39;
        goto rlab;
  rlab: u := [free](l-nth(m_12xptr, 0i), l-nth(m_12xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_13, 0i), l-nth(m_13, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_15yptr, 0i), l-nth(m_15yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_16, 0i), l-nth(m_16, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        return
};

spec "WP::<T>::assign_first"(pLft_0, pty_T0, self, x, current, proph, xm)
[[ "<&mut WP<T> as gilogic::prophecies::Ownable>::own"(#pLft_0, #pty_T0, #self, {{ #current, #proph }}) * "<T as gilogic::prophecies::Ownable>::own"(#pty_T0, #x, #xm) * <lft>(#pLft_0; true) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#self == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (xm == #xm) * (proph == #proph) * (current == #current) * (x == #x) * (self == #self) * (pty_T0 == #pty_T0) * (pLft_0 == #pLft_0) ]]
[[<observation>((#proph = {{ #xm, l-nth(#current, 1i) }}); ) * <lft>(#pLft_0; true)]]
normal

proc "WP::<T>::assign_first"(pLft_0, pty_T0, self, x) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, g_0);
        m_2x := x;
  bb0:  g_1 := m_2x;
        m_3 := g_1;
        goto bb1;
  bb1:  g_2 := m_3;
        g_4 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_3 := l-nth(g_4, 0i);
        g_6 := [load_value](l-nth(l-nth(g_3, 0i), 0i), l+ (l-nth(l-nth(g_3, 0i), 1i), {{ {{ "f", 0i, {{ "adt", "WP", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_5 := l-nth(g_6, 0i);
        u := [store_value](l-nth(g_5, 0i), l+ (l-nth(g_5, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ pty_T0 }} }} }} }}), pty_T0, g_2);
        goto bb3;
  bb3:  g_8 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_7 := l-nth(g_8, 0i);
        g_9 := [new_lft]();
        pLft_8 := l-nth(g_9, 0i);
        m_5 := g_7;
        g_10 := m_5;
        g_11 := "<&mut WP<T> as gilogic::prophecies::Prophecised>::prophecy_auto_update"(pLft_0, pty_T0, g_10);
        m_4 := g_11;
        goto bb4;
  bb4:  g_13 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_12 := l-nth(g_13, 0i);
        g_14 := [new_lft]();
        pLft_9 := l-nth(g_14, 0i);
        m_7 := g_12;
        g_15 := m_7;
        g_16 := "<&mut WP<T> as gilogic::prophecies::Prophecised>::prophecy_resolve"(pLft_0, pty_T0, g_15);
        m_6 := g_16;
        goto bb5;
  bb5:  m_0 := {{  }};
        goto bb6;
  bb6:  g_17 := m_0;
        ret := g_17;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        return
};

spec "WP::<T>::first_mut"(pLft_a, pty_T0, self, cself, pself)
[[ "<&mut WP<T> as gilogic::prophecies::Ownable>::own"(#pLft_a, #pty_T0, #self, {{ #cself, #pself }}) * <lft>(#pLft_a; true) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#self == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (pself == #pself) * (cself == #cself) * (self == #self) * (pty_T0 == #pty_T0) * (pLft_a == #pLft_a) ]]
[["<&mut T as gilogic::prophecies::Ownable>::own"(#pLft_a, #pty_T0, ret, {{ #c, #p }}) * <observation>((l-nth(#cself, 1i) = l-nth(#pself, 1i)); ) * <lft>(#pLft_a; true)]]
normal

proc "WP::<T>::first_mut"(pLft_a, pty_T0, self) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, g_0);
        m_2 := [alloc]({{ "ref", true, pty_T0 }});
        m_3 := [alloc]({{ "ref", true, pty_T0 }});
        m_6ret := [alloc]({{ "ref", true, pty_T0 }});
  bb0:  g_2 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_3 := [new_lft]();
        pLft_12 := l-nth(g_3, 0i);
        m_5 := g_1;
        g_4 := m_5;
        apply "freeze_xy::<T>"(pLft_a,pty_T0,g_4);
        goto bb1;
  bb1:  g_6 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_5 := l-nth(g_6, 0i);
        g_8 := [load_value](l-nth(l-nth(g_5, 0i), 0i), l+ (l-nth(l-nth(g_5, 0i), 1i), {{ {{ "f", 0i, {{ "adt", "WP", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_7 := l-nth(g_8, 0i);
        g_9 := [pcy_alloc]();
        g_10 := [new_lft]();
        pLft_13 := l-nth(g_10, 0i);
        u := [store_value](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }}, {{ {{ l-nth(g_7, 0i), l+ (l-nth(g_7, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ pty_T0 }} }} }} }}) }}, l-nth(g_9, 0i) }});
        g_12 := [load_value](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }}, true);
        g_11 := l-nth(g_12, 0i);
        g_13 := [new_lft]();
        pLft_11 := l-nth(g_13, 0i);
        u := [store_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }}, g_11);
        g_15 := [load_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }}, true);
        g_14 := l-nth(g_15, 0i);
        g_16 := [new_lft]();
        pLft_10 := l-nth(g_16, 0i);
        u := [store_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }}, g_14);
        g_18 := [load_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }}, true);
        g_17 := l-nth(g_18, 0i);
        g_19 := [new_lft]();
        pLft_8 := l-nth(g_19, 0i);
        m_0 := g_17;
        g_20 := m_0;
        ret := g_20;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }});
        u := [free](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }});
        u := [free](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }});
        return
};

proc "<&mut WP<T> as gilogic::prophecies::Prophecised>::prophecy_auto_update"(pLft_a, pty_T0, mut_ref) {
    sep_assert (<value>(l-nth(l-nth(mut_ref, 0i), 0i), l-nth(l-nth(mut_ref, 0i), 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #pointee) * "<WP<T> as gilogic::prophecies::Ownable>::own"(pty_T0, #pointee, #new_repr))[bind: #new_repr];
    u := [pcy_assign](l-nth(mut_ref, 1i), #new_repr);
    ret := {{  }};
    return
};

